Lib/shutil.py

```diff
@@ -1308,15 +1308,25 @@ def _access_check(fn, mode):
         path = os.environ.get("PATH", os.defpath)
     if not path:
         return None
-    path = path.split(os.pathsep)
+    if isinstance(cmd, bytes):
+        path = os.fsencode(path)
+        path = path.split(os.fsencode(os.pathsep))
+    else:
+        path = os.fsdecode(path)
+        path = path.split(os.pathsep)
 
     if sys.platform == "win32":
         # The current directory takes precedence on Windows.
-        if not os.curdir in path:
-            path.insert(0, os.curdir)
+        curdir = os.curdir
+        if isinstance(cmd, bytes):
+            curdir = os.fsencode(curdir)
+        if curdir not in path:
+            path.insert(0, curdir)
 
         # PATHEXT is necessary to check on Windows.
         pathext = os.environ.get("PATHEXT", "").split(os.pathsep)
+        if isinstance(cmd, bytes):
+            pathext = list(map(os.fsencode, pathext))
```

Comment:
Maybe use a list-comprehension instead? I'm not sure what is the best ;-)

Reply:
I think that ```list(map(os.fsencode, pathext))``` it's ok.

Reply:
I did a timeit on both and the `list(map())` is slightly faster (on my system), but not much.  If it code after this didn't (potentially) iterate over `pathext` twice, a generator would be ideal.  

Also, the following `if` clause uses:
```
    if any(cmd.lower().endswith(ext.lower()) for ext in pathext):
        files = [cmd]
```

Removing the `cmd.lower()` from the `any` would be a processing optimization, but it's unrelated to this change, so I didn't put it in.
```
    cmd_lower = cmd.lower()
    if any(cmd_lower.endswith(ext.lower()) for ext in pathext):
        files = [cmd]
```